### PR TITLE
Add gradient tokens 

### PR DESCRIPTION
**Pull request by laminesm**
### Platforms Impacted
- [x] iOS
- [ ] macOS

### Description of changes

This PR adds new gradient tokens to AliasTokens.

**Changes**

- 3 new alias color tokens `brandGradient1/2/3`. They are also added to `ColorProviding` since they're branding colors.
- 6 new global tokens in `GradientColorToken` that compose the new alias tokens 
- New alias token type `GradientTokens` as requested by design that returns an array of colors for `flair` and `tint` gradients
- Gradient token is unique in the sense that it's the only compound alias token (an alias token that returns an array of other alias tokens - branding colors in this case). So unlike other alias tokens, it has to be lazily initialized since it will need to access the instance colors to return the correct set of values.

### Binary change

Total increase: 113,600 bytes
Total decrease: -8 bytes
| File | Before | After | Delta |
|------|-------:|------:|------:|
| Total | 30,281,928 bytes | 30,395,520 bytes | 🛑 113,592 bytes |
<details>
<summary> Full breakdown </summary>

| File | Before | After | Delta |
|------|-------:|------:|------:|
| AliasTokens.o | 243,344 bytes | 278,480 bytes | ⚠️ 35,136 bytes |
| FluentTheme.o | 159,072 bytes | 181,976 bytes | ⚠️ 22,904 bytes |
| __.SYMDEF | 4,709,632 bytes | 4,726,840 bytes | ⚠️ 17,208 bytes |
| GlobalTokens.o | 543,344 bytes | 557,416 bytes | ⚠️ 14,072 bytes |
| FluentTheme+Tokens.o | 61,576 bytes | 74,256 bytes | ⚠️ 12,680 bytes |
| CommandBarCommandGroupsView.o | 201,760 bytes | 206,720 bytes | ⚠️ 4,960 bytes |
| BadgeField.o | 554,168 bytes | 557,824 bytes | ⚠️ 3,656 bytes |
| ColorProviding.o | 38,400 bytes | 40,736 bytes | ⚠️ 2,336 bytes |
| FocusRingView.o | 804,744 bytes | 805,336 bytes | ⚠️ 592 bytes |
| BadgeLabelTokenSet.o | 55,488 bytes | 55,512 bytes | ⚠️ 24 bytes |
| CommandBarTokenSet.o | 63,856 bytes | 63,872 bytes | ⚠️ 16 bytes |
| NavigationBarTokenSet.o | 73,592 bytes | 73,600 bytes | ⚠️ 8 bytes |
| DateTimePickerViewComponent.o | 136,096 bytes | 136,104 bytes | ⚠️ 8 bytes |
| DateTimePickerController.o | 140,104 bytes | 140,096 bytes | 🎉 -8 bytes |
</details>

### Verification

Gradient overrides and theme overrides work properly.

<details>
<summary>Visual Verification</summary>

| Pic 1                                       | Pic 2                                      |
|----------------------------------------------|--------------------------------------------|
| blue_light | blue_dark |
| green_light | green_dark |
| flait | tint |
</details>

### Pull request checklist

This PR has considered:
- [x] Light and Dark appearances
- [ ] iOS supported versions (all major versions greater than or equal current target deployment version)
- [ ] VoiceOver and Keyboard Accessibility
- [ ] Internationalization and Right to Left layouts
- [ ] Different resolutions (1x, 2x, 3x)
- [ ] Size classes and window sizes (iPhone vs iPad, notched devices, multitasking, different window sizes, etc)
- [ ] iPad [Pointer interaction](https://developer.apple.com/documentation/uikit/pointer_interactions)
- [ ] [SwiftUI](https://developer.apple.com/tutorials/swiftui) consumption (validation or new demo scenarios needed)
- [ ] Objective-C exposure (provide it only if needed)
 ###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/fluentui-apple/pull/1777)

### CODE REVIEW

File: ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift [hunk 1]
```diff
@@ -71,6 +71,36 @@ public class GlobalTokens: NSObject {
         }
     }
 
+    // MARK: - GradientColor
+
+    @objc(MSFGradientColorToken)
+    public enum GradientColorToken: Int, TokenSetKey {
```
Comment (huwilkes) [inline ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift:77]: I think I'd prefer these to be in the `BrandColorToken` section, so that anything for a brand alias token is coming from there, but I'm not too picky.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoColorThemes.swift [hunk 2]
```diff
@@ -208,4 +223,19 @@ class DemoColorGreenTheme: NSObject, ColorProviding {
...
                                                   dark: ColorValue(0x60BD82)))
     }
+
+    var brandGradient1: UIColor {
+        return UIColor(dynamicColor: DynamicColor(light: ColorValue(0x107C41),
+                                                  dark: ColorValue(0x10893C)))
+    }
+    
+    var brandGradient2: UIColor {
+        return UIColor(dynamicColor: DynamicColor(light: ColorValue(0xDCF51D),
```
Comment (mischreiber) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoColorThemes.swift:233]: Nit: if the value is the same for light and dark then there's no need to use a `DynamicColor`. Same with `brandGradient3`.

File: ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift [hunk 3]
```diff
@@ -447,6 +482,18 @@ extension AliasTokens {
         }
     }
 
+    private func defaultGradientColors(_ token: GradientTokens) -> [UIColor] {
+        switch token {
+        case .flair:
+            return [UIColor(dynamicColor: colors[.brandGradient1]),
+                    UIColor(dynamicColor: colors[.brandGradient2]),
+                    UIColor(dynamicColor: colors[.brandGradient3])]
+        case .tint:
```
Comment (joannaquu) [inline ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:491]: No need for `brandGradient1` in `tint`?

File: ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift (hunk 1 above)
Comment (laminesm) [reply ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift:77]: The issue is that the `brandColor` function that uses `BrandColorToken` returns a UIColor and not a `ColorValue`. The token could be added to `BrandColorsToken` though to get a `ColorValue` but there is a deprecated mark over that section so I decided against adding anything in there.

File: ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift (hunk 3 above)
Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:491]: Correct, tint is just 2 and 3

File: ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift (hunk 1 above)
Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/GlobalTokens.swift:77]: Ah, there's two different brand color sections, since we haven't finished the clean up for moving to the system types.

File: ios/FluentUI.Demo/FluentUI.Demo/DemoColorThemes.swift [hunk 4]
```diff
@@ -208,4 +223,19 @@ class DemoColorGreenTheme: NSObject, ColorProviding {
         return UIColor(dynamicColor: DynamicColor(light: ColorValue(0x0F703B),
                                                   dark: ColorValue(0x60BD82)))
     }
+
```
Comment (joannaquu) [inline ios/FluentUI.Demo/FluentUI.Demo/DemoColorThemes.swift:226]: Was there a figma you pulled the green theme values from? A while back I wanted to update `ColorProviding` but I didn't have values for the green theme

File: ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift [hunk 5]
```diff
@@ -166,15 +171,36 @@ public final class AliasTokens: NSObject {
...
     init(colorOverrides: [ColorsTokens: DynamicColor]? = nil,
          shadowOverrides: [ShadowTokens: ShadowInfo]? = nil,
-         typographyOverrides: [TypographyTokens: FontInfo]? = nil) {
+         typographyOverrides: [TypographyTokens: FontInfo]? = nil,
+         gradientOverrides: [GradientTokens: [UIColor]]? = nil) {
 
         self.colors = .init(AliasTokens.defaultColors(_:), colorOverrides)
         self.shadow = .init(AliasTokens.defaultShadows(_:), shadowOverrides)
         self.typography = .init(AliasTokens.defaultTypography(_:), typographyOverrides)
+        self.gradientOverrides = gradientOverrides
```
Comment (mischreiber) [inline ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: Is this pattern intentionally different from the others above? Where we init in the lazy getter instead of during `.init`?

Comment (laminesm) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: Yes, lazy initialization was the only way I could think of to ensure `gradientColors` uses the instance's `colors` tokens in order to return the correct value. Otherwise, the [other pattern](https://github.com/microsoft/fluentui-apple/pull/1777/commits/44751ddab8dd039caa6a25a265ccd5e5c40fda5a) will always return default token values even after the branding colors have been overridden by `ColorProviding`.

Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: I think this was mentioned in the description, but since `gradientColors` returns an array of alias tokens, it's lazy so it can pull the correct overrides.

Comment (mischreiber) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: Is that true? The `ColorProviding` values should already be set in line 175 above, right?

```swift
self.colors = .init(AliasTokens.defaultColors(_:), colorOverrides)
```

Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: It was probably complaining about trying to use `self.defaultGradientColors` before calling `super.init`, and you can't get the overrides set on this theme from `AliasTokens.defaultGradientColors`.

Comment (laminesm) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: > Is that true? The `ColorProviding` values should already be set in line 175 above, right?
> 
> ```swift
> self.colors = .init(AliasTokens.defaultColors(_:), colorOverrides)
> ```

Right, but `self.colors` is an instance property and cannot be accessed in a static context to initialize `gradientColors`.

File: ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift [hunk 6]
```diff
@@ -166,15 +171,36 @@ public final class AliasTokens: NSObject {
...
+    // MARK: - Gradient Colors
+
+    @objc(MSFGradientColorAliasTokens)
+    public enum GradientTokens: Int, TokenSetKey {
+        case flair
+        case tint
+    }
+
+    @available(swift, obsoleted: 1.0, message: "This method exists for Objective-C backwards compatibility and should not be invoked from Swift. Please use the `gradientColors` property directly.")
+    @objc(aliasGradientForToken:)
```
Comment (harrieshin) [inline ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:183]: something about this objc naming doesn't sound like the return value will be array of colors

Comment (huwilkes) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:183]: Maybe we could change to `aliasGradientColorsForToken` in objc and `gradientColors` in swift?

Comment (laminesm) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:183]: Sounds good, I'll do that

File: ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift (hunk 5 above)
Comment (harrieshin) [reply ios/FluentUI/Core/Theme/Tokens/AliasTokens.swift:203]: we should write a comment on why this initialization is different because we can't make the graident initialization static func...

**Review by huwilkes — APPROVED** [review]

**Review by sophialee0416 — APPROVED** [review]

File: ios/FluentUI/Core/ColorProviding.swift [hunk 7]
```diff
@@ -30,9 +30,9 @@ public protocol ColorProviding {
     @objc var brandStroke1: UIColor { get }
     @objc var brandStroke1Pressed: UIColor { get }
     @objc var brandStroke1Selected: UIColor { get }
-    @objc var brandGradient1: UIColor { get }
-    @objc var brandGradient2: UIColor { get }
-    @objc var brandGradient3: UIColor { get }
+    @objc var brandGradient1: UIColor? { get }
+    @objc var brandGradient2: UIColor? { get }
+    @objc var brandGradient3: UIColor? { get }
```
Comment (huwilkes) [inline ios/FluentUI/Core/ColorProviding.swift:35]: Why the change to optional? I thought we had a reason for not making the other `ColorProviding` colors optional.

Comment (laminesm) [reply ios/FluentUI/Core/ColorProviding.swift:35]: Spoke with @harrieshin offline and we thought it'd be best to minimize breaking changes when updating the apps providers.
